Fill in some missing C23 math libcall handling

valadaptive opened the pull request:
The `fminimum_num`, `fmaximum_num`, and `roundeven` libcalls (and in some cases, their corresponding intrinsics) were not being handled in various components, including constant folding, value tracking, and NumericalStabilitySanitizer. I discovered this in https://github.com/llvm/llvm-project/pull/170018.

This PR fills in the gaps, allowing those libcalls to be consistently lowered to their corresponding intrinsics and constant-folded.

(EDIT: I've also changed the `round` and `ceil` libcalls to be marked as `memory(none)` to match `floor`, `nearbyint`, and `trunc`; I can move that into a separate PR, but I want to avoid stacking too many PRs since I don't have commit access and hence cannot use Graphite.)

valadaptive committed:
[InferAttrs] Mark ceil and round as memory(none)

This probably should've been done in
https://github.com/llvm/llvm-project/pull/124742, but it wasn't. The
floor, ceil, round, and trunc family of functions all behave the same.

nikic reviewed (commented):
This needs to be split into multiple PRs, at least per component.

valadaptive committed:
[InstCombine] Add regression tests for f[min][max]imum_num libcalls

Add tests corresponding to the existing fmin/fmax tests, as we'll later
begin optimizing these into intrinsics.

valadaptive committed:
[InstSimplify] Add roundeven constant-propagation tests

The libcall versions will later be optimized and constant-folded.

valadaptive committed:
[TLI] Handle roundeven libcall in hasOptimizedCodegen

This is a prerequisite to being able to turn it into an intrinsic.

valadaptive committed:
[SelectionDAG] Handle roundeven libcall in visitCall

We can now lower it to its proper instruction.

valadaptive committed:
[InferAttrs] Handle roundeven like other libm functions

It was previously missing entirely. It behaves like floor, ceil, round,
and trunc, having no side effects.

valadaptive committed:
[ConstantFolding] Handle roundeven libcalls

Basically identical to nearbyint and rint, which we already treat as
rounding to nearest with ties to even during constant folding.

valadaptive committed:
[NumericalStabilitySanitizer] Handle roundeven libcalls

Here, it's a simple matter of wiring them up.

valadaptive committed:
[TLI] Handle f[min/max]imum_num in hasOptimizedCodegen

This is a prerequisite to lowering them to intrinsics later.

valadaptive committed:
[InferAttrs] Handle f[min/max]imum_num like other libm functions

These behave like fmin and fmax.

valadaptive committed:
[ValueTracking] Handle f[min/max]imum_num in getIntrinsicForCallSite

A straightforward mapping onto the corresponding intrinsic.

valadaptive committed:
[SimplifyLibCalls] Recognize and simplify f[min/max]imumnum

Unlike fmin and fmax, these are deterministic with regards to signed
zero.

valadaptive committed:
[NumericalStabilitySanitizer] Handle f[min/max]imum_num libcalls

A simple matter of adding them to the table.

valadaptive commented:
I've split things up into much more granular commits, but I don't want to spam this repo with 13 separate pull requests, especially since I can't use Graphite.

valadaptive commented:
I've split up the PRs. There will be an unavoidable merge conflict between two of them.
